main.cpp

```diff
@@ -147,89 +130,68 @@ struct FloatType
     FloatType(float f) : value(new float(f)) {}
     ~FloatType()
     {
+        value = nullptr;
         delete value;
     } 
-    float* value = nullptr;
 
     FloatType& add(float lhs);
     FloatType& subtract(float lhs);
     FloatType& multiply(float lhs);
     FloatType& divide(float lhs);
 
-    FloatType& add(const FloatType& lhs);
-    FloatType& subtract(const FloatType& lhs);
-    FloatType& multiply(const FloatType& lhs);
-    FloatType& divide(const FloatType& lhs);
-
-    FloatType& add(const IntType& lhs);
-    FloatType& subtract(const IntType& lhs);
-    FloatType& multiply(const IntType& lhs);
-    FloatType& divide(const IntType& lhs);
+    operator float() const
+    {
+        return *value;
+    }
 
-    FloatType& add(const DoubleType& lhs);
-    FloatType& subtract(const DoubleType& lhs);
-    FloatType& multiply(const DoubleType& lhs);
-    FloatType& divide(const DoubleType& lhs);
+private:
+    float* value;
 };
 
 struct DoubleType
 {
     DoubleType(double d) : value(new double(d)) {}
     ~DoubleType()
     {
+        value = nullptr;
         delete value;
```

Comment:
you have a memory leak now
you are deleting `nullptr`, not `value`'s memory address.
- [ ] revise